File: v8-core/src/experts/pattern_measuring_objective.rs
//! pattern_measuring_objective: evaluate() port — mirror src/v8/experts/pattern_measuring_objective.py
//! bit-for-bit (PARITY_AND_IDENTITY_SPEC §3; COMPUTE_CORE_SPEC §8 S4).
//!
//! The Python family is 3 variants (head_shoulders default, double_top,
//! triangle) selected by constructor `variant_id` — separate instances. The
//! Rust dispatch surface (mod.rs `PortFn`) carries no variant parameter, so
//! this port evaluates the DEFAULT instance, which is the instance the parity
//! harness instantiates (`PatternMeasuringObjectiveExpert()`). The double_top
//! and triangle detectors are mirrored below for completeness but are not
//! reachable through the dispatch until a variant parameter exists.

use crate::experts::base::*;
use crate::simulator::Draft;
use crate::state::HistBar;

pub const PORTED: bool = true;
pub const VERSION: &str = "v1";
pub const REQUIRES: &[&str] = &["location", "volatility", "history"];
// Declared risk geometry (EXPERT_PROTOCOL §1: risk geometry is "Predeclared
// entry, stop, target, timeout and sizing inputs"; SIMULATION_TRUTH_SPEC D-028:
// R is a declared price distance). Fixed values are declared here, never
// re-literalized inside evaluate(); a structural target/stop is computed at
// the call site and overrides the key.
pub const EXPIRY_BARS: i64 = 8;

/// The variant this port evaluates: the Python default instance
/// (PatternMeasuringObjectiveExpert() -> variant_id = "head_shoulders").
const VARIANT_ID: &str = "head_shoulders";

/// Declared, LOCKED constants (D-036; "declared, never fitted").
const PT_FLANK: usize = 3;          // structure-pivot flank on the 32-bar history
const TRIANGLE_WINDOW: usize = 20;  // convergence scan window (bars)
const TRIANGLE_WIDTH_MAX: f64 = 0.03; // max consolidation width (G-26 verbatim)

/// Python `max(items, key=x[1])`: FIRST maximum on ties (Python's max returns
/// the first maximal element; Rust's max_by_key returns the last).
fn max_high(items: &[(usize, f64)]) -> (usize, f64) {
    let mut best = items[0];
    for &x in &items[1..] {
        if x.1 > best.1 {
            best = x;
        }
    }
    best
}

/// Python `min(items, key=x[1])`: FIRST minimum on ties.
fn min_low(items: &[(usize, f64)]) -> (usize, f64) {
    let mut best = items[0];
    for &x in &items[1..] {
        if x.1 < best.1 {
            best = x;
        }
    }
    best
}

/// `_pivot_highs`: bar i whose high exceeds the flank highs on both sides.
fn pivot_highs(hist: &[HistBar]) -> Vec<(usize, f64)> {
    let mut out = Vec::new();
    for i in PT_FLANK..hist.len() - PT_FLANK {
        let hi = hist[i].high;
        let left = (i - PT_FLANK..i).map(|j| hist[j].high).fold(f64::NEG_INFINITY, f64::max);
        let right = (i + 1..i + 1 + PT_FLANK).map(|j| hist[j].high).fold(f64::NEG_INFINITY, f64::max);
        if hi > left && hi > right {
            out.push((i, hi));
        }
    }
    out
}

/// `_pivot_lows`: bar i whose low is below the flank lows on both sides.
fn pivot_lows(hist: &[HistBar]) -> Vec<(usize, f64)> {
    let mut out = Vec::new();
    for i in PT_FLANK..hist.len() - PT_FLANK {
        let lo = hist[i].low;
        let left = (i - PT_FLANK..i).map(|j| hist[j].low).fold(f64::INFINITY, f64::min);
        let right = (i + 1..i + 1 + PT_FLANK).map(|j| hist[j].low).fold(f64::INFINITY, f64::min);
        if lo < left && lo < right {
            out.push((i, lo));
        }
    }
    out
}

/// `_hs_top`: (head, neckline, right_shoulder_idx) or None.
fn hs_top(hist: &[HistBar]) -> Option<(f64, f64, usize)> {
    let ph = pivot_highs(hist);
    let pl = pivot_lows(hist);
    if ph.len() < 3 || pl.len() < 2 {
        return None;
    }
    let (head_i, head) = max_high(&ph);
    let lefts: Vec<(usize, f64)> = ph.iter().copied().filter(|&(i, _)| i < head_i).collect();
    let rights: Vec<(usize, f64)> = ph.iter().copied().filter(|&(i, _)| i > head_i).collect();
    if lefts.is_empty() || rights.is_empty() {
        return None;
    }
    let (li, left) = max_high(&lefts);
    let (ri, right) = max_high(&rights);
    if !(left < head && right < head) {
        return None;
    }
    let lt: Vec<f64> = pl.iter().copied().filter(|&(i, _)| li < i && i < head_i).map(|(_, v)| v).collect();
    let rt: Vec<f64> = pl.iter().copied().filter(|&(i, _)| head_i < i && i < ri).map(|(_, v)| v).collect();
    if lt.is_empty() || rt.is_empty() {
        return None;
    }
    let neckline = lt.iter().copied().fold(f64::NEG_INFINITY, f64::max)
        .max(rt.iter().copied().fold(f64::NEG_INFINITY, f64::max));
    if neckline >= head {
        return None;
    }
    Some((head, neckline, ri))
}

/// `_hs_bottom`: (head, neckline, right_shoulder_idx) or None.
fn hs_bottom(hist: &[HistBar]) -> Option<(f64, f64, usize)> {
    let ph = pivot_highs(hist);
    let pl = pivot_lows(hist);
    if ph.len() < 2 || pl.len() < 3 {
        return None;
    }
    let (head_i, head) = min_low(&pl);
    let lefts: Vec<(usize, f64)> = pl.iter().copied().filter(|&(i, _)| i < head_i).collect();
    let rights: Vec<(usize, f64)> = pl.iter().copied().filter(|&(i, _)| i > head_i).collect();
    if lefts.is_empty() || rights.is_empty() {
        return None;
    }
    let (li, left) = min_low(&lefts);
    let (ri, right) = min_low(&rights);
    if !(left > head && right > head) {
        return None;
    }
    let lp: Vec<f64> = ph.iter().copied().filter(|&(i, _)| li < i && i < head_i).map(|(_, v)| v).collect();
    let rp: Vec<f64> = ph.iter().copied().filter(|&(i, _)| head_i < i && i < ri).map(|(_, v)| v).collect();
    if lp.is_empty() || rp.is_empty() {
        return None;
    }
    let neckline = lp.iter().copied().fold(f64::INFINITY, f64::min)
        .min(rp.iter().copied().fold(f64::INFINITY, f64::min));
    if neckline <= head {
        return None;
    }
    Some((head, neckline, ri))
}

/// `_head_shoulders`: neckline break on the current close (no retest — entry
/// on the breakout), target = head-to-neckline height. Returns
/// (direction, level, stop_price, height, anchor).
fn head_shoulders(hist: &[HistBar]) -> Option<(String, f64, f64, f64, String)> {
    let close = hist[hist.len() - 1].close;
    if let Some((head, neckline, ri)) = hs_top(hist) {
        if close < neckline {
            let pred = |j: usize, bar: &HistBar| j >= ri && bar.close < neckline;
            let anchor = find_setup_anchor(hist, &pred);
            return Some(("SHORT".to_string(), neckline, head, head - neckline, anchor));
        }
    }
    if let Some((head, neckline, ri)) = hs_bottom(hist) {
        if close > neckline {
            let pred = |j: usize, bar: &HistBar| j >= ri && bar.close > neckline;
            let anchor = find_setup_anchor(hist, &pred);
            return Some(("LONG".to_string(), neckline, head, neckline - head, anchor));
        }
    }
    None
}

/// `_double`: validation-level break on the current close; target = the 1:1
/// validation-to-extreme projection. SHORT on the trough between the two most
/// recent pivot highs, LONG on the peak between the two most recent pivot lows.
#[allow(dead_code)] // mirrored; not reachable until a variant dispatch exists
fn double_detector(hist: &[HistBar]) -> Option<(String, f64, f64, f64, String)> {
    let close = hist[hist.len() - 1].close;
    let ph = pivot_highs(hist);
    if ph.len() >= 2 {
        let (i2, h2) = ph[ph.len() - 2];
        let (i1, h1) = ph[ph.len() - 1];
        let level = (i2 + 1..i1).map(|j| hist[j].low).fold(f64::INFINITY, f64::min);
        if h1 > level && h2 > level && close < level {
            let stop = h1.max(h2);
            let pred = |j: usize, bar: &HistBar| j >= i1 && bar.close < level;
            let anchor = find_setup_anchor(hist, &pred);
            return Some(("SHORT".to_string(), level, stop, stop - level, anchor));
        }
    }
    let pl = pivot_lows(hist);
    if pl.len() >= 2 {
        let (i2, v2) = pl[pl.len() - 2];
        let (i1, v1) = pl[pl.len() - 1];
        let level = (i2 + 1..i1).map(|j| hist[j].high).fold(f64::NEG_INFINITY, f64::max);
        if v1 < level && v2 < level && close > level {
            let stop = v1.min(v2);
            let pred = |j: usize, bar: &HistBar| j >= i1 && bar.close > level;
            let anchor = find_setup_anchor(hist, &pred);
            return Some(("LONG".to_string(), level, stop, level - stop, anchor));
        }
    }
    None
}

/// `_triangle_structure`: >= 2 pivot highs declining and >= 2 pivot lows
/// rising inside the trailing consolidation window (excluding the current bar).
#[allow(dead_code)] // mirrored; not reachable until a variant dispatch exists
fn triangle_structure(hist: &[HistBar]) -> bool {
    let n = hist.len();
    let lo = n.saturating_sub(1 + TRIANGLE_WINDOW);
    let mut ph: Vec<(usize, f64)> =
        pivot_highs(hist).into_iter().filter(|&(i, _)| lo <= i && i < n - 1).collect();
    let mut pl: Vec<(usize, f64)> =
        pivot_lows(hist).into_iter().filter(|&(i, _)| lo <= i && i < n - 1).collect();
    // Python `sorted((i, h) for ...)`: tuple order is index order (indices are
    // distinct), so sorting by the pivot index alone is equivalent.
    ph.sort_by_key(|&(i, _)| i);
    pl.sort_by_key(|&(i, _)| i);
    if ph.len() < 2 || pl.len() < 2 {
        return false;
    }
    ph[0].1 > ph[ph.len() - 1].1 && pl[0].1 < pl[pl.len() - 1].1
}

/// `_triangle`: close beyond the narrow consolidation range (breakout from a
/// converging range); target = 1:1 of the 20-bar range height.
#[allow(dead_code)] // mirrored; not reachable until a variant dispatch exists
fn triangle_detector(fm: &FeatMap) -> Option<(String, f64, f64, f64, String)> {
    let hist = &fm.history;
    let n = hist.len();
    let cr = fm.features.get("consolidation_range").map(|f| f.value.clone());
    let rh = fm.value("range_height_20");
    let cr = match cr {
        Some(c) => c,
        None => return None,
    };
    let rh = match rh {
        Some(v) => v,
        None => return None,
    };
    let h_ref = cr[0].as_f64()?;
    let l_ref = cr[1].as_f64()?;
    let width_ratio = cr[2].as_f64()?;
    if width_ratio > TRIANGLE_WIDTH_MAX || !triangle_structure(hist) {
        return None;
    }
    if n < 2 {
        return None;
    }
    let prev_close = hist[n - 2].close;
    if !(l_ref <= prev_close && prev_close <= h_ref) {
        return None;
    }
    let close = fm.value("close")?;
    if close > h_ref {
        let pred = |_j: usize, bar: &HistBar| bar.close > h_ref;
        let anchor = find_setup_anchor(hist, &pred);
        return Some(("LONG".to_string(), h_ref, l_ref, rh, anchor));
    }
    if close < l_ref {
        let pred = |_j: usize, bar: &HistBar| bar.close < l_ref;
        let anchor = find_setup_anchor(hist, &pred);
        return Some(("SHORT".to_string(), l_ref, h_ref, rh, anchor));
    }
    None
}

/// evaluate() — mirrors the Python default instance (variant head_shoulders).
pub fn pattern_measuring_objective(fm: &FeatMap, expert_id: &str, version: &str) -> ExpertEval {
    let sym = fm.symbol;
    let close = match fm.value("close") { Some(v) => v, None => return no_habitat(expert_id, version, fm.as_of) };
    let atr = match fm.value("atr") { Some(v) => v, None => return no_habitat(expert_id, version, fm.as_of) };
    // `_need`: the seven declared features must all be present
    // (consolidation_range is structured — presence, not value read).
    for k in ["history", "window_high_20", "window_low_20",
              "range_height_20", "consolidation_range"] {
        if !fm.features.contains_key(k) {
            return no_habitat(expert_id, version, fm.as_of);
        }
    }
    if fm.history.is_empty() {
        return no_habitat(expert_id, version, fm.as_of);
    }
    let n = fm.history.len();
    if n < 2 * PT_FLANK + 1 {
        return no_habitat(expert_id, version, fm.as_of);
    }
    let hit = match VARIANT_ID {
        "head_shoulders" => head_shoulders(&fm.history),
        "double_top" => double_detector(&fm.history),
        _ => triangle_detector(fm),
    };
    let (direction, level, stop_price, height, anchor) = match hit {
        Some(h) => h,
        None => return no_setup(expert_id, version, fm.as_of),
    };
    let stop_r = if direction == "LONG" { (close - stop_price) / atr } else { (stop_price - close) / atr };
    let target_r = height / atr;
    if stop_r <= 0.0 || target_r <= 0.0 {
        return no_setup(expert_id, version, fm.as_of);
    }
    let mut entries = vec![
        ("entry", serde_json::json!("NEXT_BAR_CLOSE")),
        ("target_r", serde_json::json!(target_r)),
        ("stop_r", serde_json::json!(stop_r)),
        ("expiry_bars", serde_json::json!(EXPIRY_BARS)),
        ("atr_ref", serde_json::json!(atr)),
        ("variant", serde_json::json!(VARIANT_ID)),
        ("level_ref", serde_json::json!(level)),
        ("stop_ref", serde_json::json!(stop_price)),
    ];
    if direction == "LONG" {
        entries.push(("prior_low_ref", serde_json::json!(stop_price)));
    } else {
        entries.push(("prior_high_ref", serde_json::json!(stop_price)));
    }
    let draft = Draft {
        direction: direction.to_string(),
        birth_time: fm.as_of,
        risk_geometry: geom(entries),
    };
    let fingerprint = format!("{sym}:{VARIANT_ID}:{direction}:{close:.6}:{level:.6}:{stop_price:.6}");
    candidate(expert_id, version, fm.as_of, draft, anchor, fingerprint)
}

File: src/v8/experts/pattern_measuring_objective.py
"""Geometric-pattern-breakout behavior family (`pattern_measuring_objective`).

Hypothesis (mechanism `geometric_pattern_breakout`; Ch13.2 p499-501): a
completed geometric pattern whose boundary is broken by a CLOSE is a
breakout, and the book's measuring-objective doctrine fixes the target at the
1:1 projection of the pattern's vertical height — "a completed pattern that
achieves its minimum 1:1 objective avoids future pattern-failure
classification" (Ch13.3 p509). Targets become pattern-derived price
distances expressed in R with the 14-bar ATR unit (D-028).

Variants (D-044; one per pattern, all sharing the measuring-objective
mechanism):
  head_shoulders  - head-and-shoulders: neckline break (top -> SHORT, bottom
                    -> LONG); target = 1:1 head-to-neckline height; stop
                    beyond the head.
  double_top      - double top/bottom: validation-level break (the trough
                    between two peaks for a top, the peak between two troughs
                    for a bottom); target = 1:1 validation-to-extreme height;
                    stop beyond the higher peak / lower trough.
  triangle        - symmetrical triangle: close beyond the consolidation
                    range after a narrow, converging range; target = 1:1 of
                    the 20-bar range height; stop behind the opposite range
                    bound.

The book's flag/pennant, wedge, rounding and triple-top patterns are not
implemented in this family: flags/pennants need pole-parallelogram geometry
that the 32-bar history pin (O-020) cannot express, and the card lists them
under the same mechanism with a shared `variants_evaluated` for a future
revision. Structure pivots use a small declared flank because the global
history window is 32 bars (O-020).

The pattern line (neckline / validation level / range bound) is FROZEN at
detection (the prior_high_ref pattern); `still_valid` keeps the thesis alive
while the close stays beyond it — a close back through says the pattern
failed (Ch13.3: "minimum objective not met before price reverses through the
breakout level").
"""
from __future__ import annotations

from ..schema import MarketState, CandidateDraft, ExpertEvaluation
from .base import Expert


class PatternMeasuringObjectiveExpert(Expert):
    """Geometric pattern breakout with a 1:1 measuring-objective target."""
    expert_id = 'pattern_measuring_objective'
    version = 'v1'
    mechanism_family_id = 'geometric_pattern_breakout'
    behavior_family_id = 'geometric_pattern_breakout'
    variant_id = 'head_shoulders'
    # target_r/stop_r are structural: the 1:1 pattern-height measuring
    # objective and the frozen pattern stop in R (D-028), computed in
    # evaluate().
    target_r = None
    stop_r = None
    # D-044: every implemented variant (losers included); the reported
    # variant_id is a member. D-046: every threshold/lookback below is a
    # declared constant frozen pre-window, so the search universe equals the
    # evaluated set.
    variants_evaluated = ('head_shoulders', 'double_top', 'triangle')
    search_universe_size = 3
    requires = ('location', 'volatility', 'history')

    # Declared, LOCKED constants (D-036 pattern: "declared, never fitted").
    PT_FLANK = 3                 # structure-pivot flank on the 32-bar history
    TRIANGLE_WINDOW = 20         # convergence scan window (bars)
    TRIANGLE_WIDTH_MAX = 0.03    # max consolidation width (G-26 verbatim)

    def __init__(self, variant_id: str | None = None):
        if variant_id is not None:
            if variant_id not in self.variants_evaluated:
                raise ValueError(
                    f'{self.expert_id}: unknown variant {variant_id!r} '
                    f'(variants_evaluated={list(self.variants_evaluated)})')
            self.variant_id = variant_id

    # --- pivot lattice over the history window -------------------------------

    def _pivot_highs(self):
        out = []
        flank = self.PT_FLANK
        hist = self._hist
        for i in range(flank, len(hist) - flank):
            hi = float(hist[i][2])
            if hi > max(float(hist[j][2]) for j in range(i - flank, i)) \
                    and hi > max(float(hist[j][2])
                                 for j in range(i + 1, i + 1 + flank)):
                out.append((i, hi))
        return out

    def _pivot_lows(self):
        out = []
        flank = self.PT_FLANK
        hist = self._hist
        for i in range(flank, len(hist) - flank):
            lo = float(hist[i][3])
            if lo < min(float(hist[j][3]) for j in range(i - flank, i)) \
                    and lo < min(float(hist[j][3])
                                 for j in range(i + 1, i + 1 + flank)):
                out.append((i, lo))
        return out

    # --- head-and-shoulders structure ----------------------------------------

    def _hs_top(self) -> tuple | None:
        """(head, neckline, right_shoulder_idx) or None."""
        ph = self._pivot_highs()
        pl = self._pivot_lows()
        if len(ph) < 3 or len(pl) < 2:
            return None
        head_i, head = max(ph, key=lambda x: x[1])
        lefts = [(i, h) for i, h in ph if i < head_i]
        rights = [(i, h) for i, h in ph if i > head_i]
        if not lefts or not rights:
            return None
        li, left = max(lefts, key=lambda x: x[1])
        ri, right = max(rights, key=lambda x: x[1])
        if not (left < head and right < head):
            return None
        lt = [v for i, v in pl if li < i < head_i]
        rt = [v for i, v in pl if head_i < i < ri]
        if not lt or not rt:
            return None
        neckline = max(max(lt), max(rt))
        if neckline >= head:
            return None
        return head, neckline, ri

    def _hs_bottom(self) -> tuple | None:
        """(head, neckline, right_shoulder_idx) or None."""
        ph = self._pivot_highs()
        pl = self._pivot_lows()
        if len(ph) < 2 or len(pl) < 3:
            return None
        head_i, head = min(pl, key=lambda x: x[1])
        lefts = [(i, v) for i, v in pl if i < head_i]
        rights = [(i, v) for i, v in pl if i > head_i]
        if not lefts or not rights:
            return None
        li, left = min(lefts, key=lambda x: x[1])
        ri, right = min(rights, key=lambda x: x[1])
        if not (left > head and right > head):
            return None
        lp = [v for i, v in ph if li < i < head_i]
        rp = [v for i, v in ph if head_i < i < ri]
        if not lp or not rp:
            return None
        neckline = min(min(lp), min(rp))
        if neckline <= head:
            return None
        return head, neckline, ri

    # --- variant detection ----------------------------------------------------

    def _head_shoulders(self, n: int) -> tuple | None:
        """Neckline break on the current close (no retest — entry on the
        breakout), target = head-to-neckline height."""
        close = float(self._hist[n - 1][4])
        top = self._hs_top()
        if top is not None:
            head, neckline, ri = top
            if close < neckline:
                pred = lambda j, bar: j >= ri and float(bar[4]) < neckline
                anchor = self.find_setup_anchor(self._hist, pred)
                return 'SHORT', neckline, head, head - neckline, anchor
        bottom = self._hs_bottom()
        if bottom is not None:
            head, neckline, ri = bottom
            if close > neckline:
                pred = lambda j, bar: j >= ri and float(bar[4]) > neckline
                anchor = self.find_setup_anchor(self._hist, pred)
                return 'LONG', neckline, head, neckline - head, anchor
        return None

    def _double(self, n: int) -> tuple | None:
        """Validation-level break on the current close; target = the 1:1
        validation-to-extreme projection."""
        close = float(self._hist[n - 1][4])
        ph = self._pivot_highs()
        if len(ph) >= 2:
            (i2, h2), (i1, h1) = ph[-2], ph[-1]
            level = min(float(self._hist[j][3]) for j in range(i2 + 1, i1))
            if h1 > level and h2 > level and close < level:
                stop = max(h1, h2)
                pred = lambda j, bar: j >= i1 and float(bar[4]) < level
                anchor = self.find_setup_anchor(self._hist, pred)
                return 'SHORT', level, stop, stop - level, anchor
        pl = self._pivot_lows()
        if len(pl) >= 2:
            (i2, v2), (i1, v1) = pl[-2], pl[-1]
            level = max(float(self._hist[j][2]) for j in range(i2 + 1, i1))
            if v1 < level and v2 < level and close > level:
                stop = min(v1, v2)
                pred = lambda j, bar: j >= i1 and float(bar[4]) > level
                anchor = self.find_setup_anchor(self._hist, pred)
                return 'LONG', level, stop, level - stop, anchor
        return None

    def _triangle_structure(self) -> bool:
        """Symmetrical-triangle convergence: at least 2 pivot highs DECLINING
        and 2 pivot lows RISING inside the trailing consolidation window
        (excluding the current bar)."""
        n = len(self._hist)
        lo = max(0, n - 1 - self.TRIANGLE_WINDOW)
        ph = sorted((i, h) for i, h in self._pivot_highs() if lo <= i < n - 1)
        pl = sorted((i, v) for i, v in self._pivot_lows() if lo <= i < n - 1)
        if len(ph) < 2 or len(pl) < 2:
            return False
        return ph[0][1] > ph[-1][1] and pl[0][1] < pl[-1][1]

    def _triangle(self, f: dict, n: int) -> tuple | None:
        """Close beyond the narrow consolidation range (breakout from a
        converging range); target = 1:1 of the range height."""
        sym = self._sym
        cr = f.get(f'{sym}.consolidation_range')
        rh = f.get(f'{sym}.range_height_20')
        if cr is None or cr.value is None or rh is None or rh.value is None:
            return None
        h_ref, l_ref, width_ratio, _active = cr.value
        if width_ratio > self.TRIANGLE_WIDTH_MAX or not self._triangle_structure():
            return None
        prev_close = float(self._hist[n - 2][4]) if n >= 2 else None
        if prev_close is None or not (l_ref <= prev_close <= h_ref):
            return None                       # no consolidation right before
        close = float(f[f'{sym}.close'].value)
        height = float(rh.value)
        if close > h_ref:
            pred = lambda j, bar: float(bar[4]) > h_ref
            anchor = self.find_setup_anchor(self._hist, pred)
            return 'LONG', h_ref, l_ref, height, anchor
        if close < l_ref:
            pred = lambda j, bar: float(bar[4]) < l_ref
            anchor = self.find_setup_anchor(self._hist, pred)
            return 'SHORT', l_ref, h_ref, height, anchor
        return None

    # --- evaluate / still_valid -----------------------------------------------

    def evaluate(self, state: MarketState) -> ExpertEvaluation:
        t = state.as_of
        sym = state.universe[0]
        need = [f'{sym}.close', f'{sym}.atr', f'{sym}.history',
                f'{sym}.window_high_20', f'{sym}.window_low_20',
                f'{sym}.range_height_20', f'{sym}.consolidation_range']
        if not self._need(state, need):
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        f = state.features
        close = float(f[f'{sym}.close'].value)
        atr = f[f'{sym}.atr'].value
        hist_value = f[f'{sym}.history'].value
        if not isinstance(hist_value, (tuple, list)) or not hist_value or atr is None:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        self._hist = tuple(hist_value)
        self._sym = sym
        n = len(self._hist)
        if n < 2 * self.PT_FLANK + 1:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        if self.variant_id == 'head_shoulders':
            hit = self._head_shoulders(n)
        elif self.variant_id == 'double_top':
            hit = self._double(n)
        else:
            hit = self._triangle(f, n)
        if hit is None:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_SETUP', t)
        direction, level, stop_price, height, anchor = hit
        if direction == 'LONG':
            stop_r = (close - stop_price) / atr
        else:
            stop_r = (stop_price - close) / atr
        target_r = height / atr
        if stop_r <= 0 or target_r <= 0:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_SETUP', t)
        geometry = self.declared_geometry()
        geometry.update({'target_r': target_r, 'stop_r': stop_r,
                         'atr_ref': atr, 'variant': self.variant_id,
                         'level_ref': level, 'stop_ref': stop_price})
        if direction == 'LONG':
            geometry['prior_low_ref'] = stop_price
        else:
            geometry['prior_high_ref'] = stop_price
        draft = CandidateDraft(
            expert_id=self.expert_id, expert_version=self.version,
            instrument=sym, direction=direction,
            setup_fingerprint=(f'{sym}:{self.variant_id}:{direction}:'
                               f'{close:.6f}:{level:.6f}:{stop_price:.6f}'),
            risk_geometry=geometry,
            birth_time=t, setup_anchor_event_id=anchor)
        return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                'APPLICABLE', 'CANDIDATE', t, draft)

    def still_valid(self, state: MarketState, draft: CandidateDraft) -> bool:
        """The thesis is "the pattern's completion line holds": a close back
        through the FROZEN line says the breakout failed and the 1:1
        objective is not in force."""
        sym = draft.instrument
        f = state.features
        close = f.get(f'{sym}.close')
        if close is None or close.value is None:
            return True          # unobservable: fail open, price still governs
        ref = draft.risk_geometry.get('level_ref')
        if ref is None:
            return True
        if draft.direction == 'LONG':
            return float(close.value) > float(ref)
        return float(close.value) < float(ref)
